tests/xharness/TestReporter.cs

```diff
@@ -0,0 +1,490 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Xml;
+using System.Runtime.Serialization.Json;
+using Xharness.Execution;
+using Xharness.Listeners;
+using Xharness.Logging;
+using Xharness.Utilities;
+
+namespace Xharness {
+
+	public class TestReporterFactory : ITestReporterFactory {
+
+		public ITestReporter Create (IAppRunner appRunner, string device, ISimpleListener simpleListener, ILog log, ICrashSnapshotReporter crashReports)
+			=> new TestReporter (appRunner, device, simpleListener, log, crashReports);
+	}
+
+	// main class that gets the result of an executed test application, parses the results and provides information
+	// about the success or failure of the execution.
+	public class TestReporter : ITestReporter {
+
+		const string timeoutMessage = "Test run timed out after {0} minute(s).";
+		const string completionMessage = "Test run completed";
+		const string failureMessage = "Test run failed";
+
+		public TimeSpan Timeout { get; private set; }
+		public IProcessManager ProcessManager { get; set; } = new ProcessManager ();
+		public IResultParser ResultParser { get; set; } = new XmlResultParser ();
+		public ILog CallbackLog { get; private set; }
+
+		public bool? Success { get; private set; }
+		public Stopwatch TimeoutWatch { get; private set; }
+		public CancellationToken CancellationToken => cancellationTokenSource.Token;
+
+		public bool ResultsUseXml => runner.XmlJargon != XmlResultJargon.Missing;
+
+		readonly IAppRunner runner;
+		readonly ISimpleListener listener;
+		readonly ILogs crashLogs;
+		readonly ILog runLog;
+		readonly ICrashSnapshotReporter crashReporter;
+		readonly string deviceName;
+		bool waitedForExit = true;
+		bool launchFailure;
+		bool isSimulatorTest;
+		bool timedout;
+
+		readonly CancellationTokenSource cancellationTokenSource = new CancellationTokenSource ();
+
+		public TestReporter (IAppRunner appRunner, string device, ISimpleListener simpleListener, ILog log, ICrashSnapshotReporter crashReports)
+		{
+			runner = appRunner ?? throw new ArgumentNullException (nameof (appRunner));
+			deviceName = device; // can be null on simulators 
+			listener = simpleListener ?? throw new ArgumentNullException (nameof (simpleListener));
+			runLog = log ?? throw new ArgumentNullException (nameof (log));
+			crashLogs = new Logs (runner.Logs.Directory);
+			crashReporter = crashReports ?? throw new ArgumentNullException (nameof (crashReports));
+			Timeout = runner.GetNewTimeout ();
+			TimeoutWatch  = new Stopwatch ();
+			CallbackLog = new CallbackLog ((line) => {
```

Comment:
Same logic as the one found in RunAsync. Moved here since we are interested of the wait in this class.